util/stmtsummary/statement_summary.go

```diff
@@ -414,228 +479,311 @@ func newStmtSummaryByDigest(sei *StmtExecInfo, beginTime int64) *stmtSummaryByDi
 	tableNames := buffer.String()
 
 	ssbd := &stmtSummaryByDigest{
-		beginTime:     beginTime,
 		schemaName:    sei.SchemaName,
 		digest:        sei.Digest,
 		stmtType:      strings.ToLower(sei.StmtCtx.StmtType),
 		normalizedSQL: normalizedSQL,
 		tableNames:    tableNames,
-		indexNames:    strings.Join(sei.StmtCtx.IndexNames, ","),
-		minLatency:    sei.TotalLatency,
-		backoffTypes:  make(map[fmt.Stringer]int),
-		firstSeen:     sei.StartTime,
-		lastSeen:      sei.StartTime,
+		history:       list.New(),
 	}
-	ssbd.add(sei)
+	ssbd.add(sei, beginTime, historySize)
 	return ssbd
 }
 
-func (ssbd *stmtSummaryByDigest) add(sei *StmtExecInfo) {
+func (ssbd *stmtSummaryByDigest) add(sei *StmtExecInfo, beginTime int64, historySize int) {
+	// Enclose this block in a function to ensure the lock will always be released.
+	ssElement, ok := func() (*stmtSummaryByDigestElement, bool) {
+		ssbd.Lock()
+		defer ssbd.Unlock()
+
+		var ssElement *stmtSummaryByDigestElement
+		ok := false
+		if ssbd.history.Len() > 0 {
+			lastElement := ssbd.history.Back().Value.(*stmtSummaryByDigestElement)
+			if lastElement.beginTime >= beginTime {
+				ssElement = lastElement
+				ok = true
+			}
+		}
+		if !ok {
+			ssElement = newStmtSummaryByDigestElement(sei, beginTime)
+			ssbd.history.PushBack(ssElement)
+		}
+
+		// `historySize` might be modified, so check expiration every time.
+		if historySize == 0 {
+			historySize = 1
```

Comment:
This assignment looks weird. Why reassign `historySize` here?

Reply:
If `historySize` is 0, displayed history summaries in `events_statements_summary_by_digest_history` should be empty, but summary cannot be cleared, because table `events_statements_summary_by_digest` still needs it.
Alternatively, I can code `for ssbd.history.Len() > historySize && ssbd.history.Len() > 1 {...}`. But it looks also weird.

Reply:
Addressed.